### ISSUE TITLE
ESLint Pre-Commit hook failing on installation due to peer conflict with mocha

### ISSUE THREAD

**Issue by CAM-Gerlach** [eslint/eslint#18956]
## TL;DR

After an all too lengthy investigation, mostly due to me not being a NodeJS or NPM user but rather using ESLint to check frontend JS scripts as part of much larger non-JS projects, this appears to be solved by #18945 once a release is cut with it (which I'm inferring may not be until #18951 ). I evidently missed it despite a lot of googling as it had evidently not yet been indexed by the search engines I used, and I'd been mainly searching issues rather than PRs and it was not documented in the former, only fixed in the latter; I only ended up finding it when I happened to see it in `git blame`.

However, given I'd written nearly all of it up already, I figured I'd go ahead and post it to properly document the problem and how it was narrowed down for posterity and anyone else who may come run it in the meantime. And more practically, I'd like to know and document if there's a better workaround than what I came up with (downgrading the ESLint version to 9.9.0), since I tried to just explicitly pin the dep version before the earlier dependency version without the incompatible peer dep, but NPM still pulls the latest version and produces the same error.

## Background

Hi! I'm a core dev of the [Spyder FOSS scientific Python IDE](https://www.spyder-ide.org) (as well as the Python programming language). Within the Spyder community, we happily use ESLint via its official Pre-Commit hook across a growing number of projects in [our organization](https://github.com/spyder-ide), primarily to check and lint JS frontend scripts that are part of larger, non-JS projects.

We recently upgraded from ESLint v8.52.0 to v9.11.0 in spyder-ide/spyder-docs#383, soon to be mirrored org-wide, and after a somewhat painful process migrating to a flat JS config file (mostly down to not being a Node developer or working on a Node project), everything worked fine in that PR, as did spyder-ide/spyder-docs#384 (using the same cached Pre-Commit environment). 

## Issue summary

After submitting a PR touching an unrelated hook in the Pre-Commit config file (which invalidated the GitHub Actions cache of the pre-commit env), the ESlint hook suddenly [started to fail to install](https://github.com/spyder-ide/spyder-docs/actions/runs/11027060843/job/30624671687) at all in spyder-ide/spyder-docs#385 due to a peer dependency conflict with ``mocha``, causing an immediate exit with a critical error, even after re-running the job a full day later. This error was also [reproduced on `master`](https://github.com/spyder-ide/spyder-docs/actions/runs/11046390213/job/30685628175) after clearing the Pre-Commit cache, where the same commit/ESLint hook/version previously passed, and locally with just `rev: 9.11.1` for the hook with no `additional_depenencies` or other config.

I've traced the issue back to what looks to me like a dependency conflict between two of ESlint's dev dependencies, but as I'm not a NodeJS developer or NPM user, much less an ESLint maintainer, I'm not sure how best to address this.

## Error message

`pre-commit run --all eslint` exits with

```
[INFO] Initializing environment for https://github.com/eslint/eslint.
[INFO] Installing environment for https://github.com/eslint/eslint.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
An unexpected error has occurred: CalledProcessError: command: ('/home/runner/.cache/pre-commit/repo1_7bgc4d/node_env-system/bin/node', '/usr/local/bin/npm', 'install', '--include=dev', '--include=prod', '--ignore-prepublish', '--no-progress', '--no-save')
return code: 1
stdout: (none)
stderr:
    npm error code ERESOLVE
    npm error ERESOLVE unable to resolve dependency tree
    npm error
    npm error While resolving: eslint@9.11.0
    npm error Found: mocha@8.4.0
    npm error node_modules/mocha
    npm error   dev mocha@"^8.3.2" from the root project
    npm error
    npm error Could not resolve dependency:
    npm error peer mocha@"^10.7.3" from @wdio/browser-runner@9.1.0
    npm error node_modules/@wdio/browser-runner
    npm error   dev @wdio/browser-runner@"^9.0.5" from the root project
    npm error
    npm error Fix the upstream dependency conflict, or retry
    npm error this command with --force or --legacy-peer-deps
    npm error to accept an incorrect (and potentially broken) dependency resolution.
```

See [full run output](https://github.com/spyder-ide/spyder-docs/actions/runs/11027060843/job/30624671687)

## Reproducer

I was able to reproduce an identical error locally (on Windows) with just the latest version of ESLint and Pre-Commit and no plugins or custom config; i.e. just

```yaml
# .pre-commit-config.yaml
repos:
- repo: https://github.com/eslint/eslint
  rev: v9.11.1
  hooks:
  - id: eslint
```

To reproduce, save the above to `.pre-commit-config.yaml` at the root of any repo with any (non-.gitignored) `.js` file and run `pre-commit run --all eslint` to trigger the error.

## Versions and timeframe affected

The [last non-cached job that succeeded](https://github.com/spyder-ide/spyder-docs/actions/runs/10978597236) started at 2024-09-22 05:29 UTC, while the [first fresh job that failed](https://github.com/spyder-ide/spyder-docs/actions/runs/11027060843) kicked off at 2024-09-25 05:46 UTC. Therefore, the problematic dependency update must presumably have occurred between those two times. Presuming this was [`browser-runner`](https://www.npmjs.com/package/@wdio/browser-runner), the only update at or near that range was [9.1.0](https://github.com/webdriverio/webdriverio/releases/tag/v9.1.0).

After manual bisection testing with Eslint/hook versions 9.0.0 through 9.11.1, versions >=9.9.1 were affected (9.9.1, 9.10.0, 9.11.0 and 9.11.1 tested), while versions prior (9.0.0, 9.5.0, 9.8.0, 9.9.1 tested) did not.

## Probable immediate cause

Browser-Runner v9.1.0 [added the following to its `package.json`](https://github.com/webdriverio/webdriverio/compare/v9.0.9...v9.1.0#diff-e022b199cd426448e3feb32597cfa9278dcf396d95c9968ca887fba33a9a6774):

```json
  "peerDependencies": {
    "mocha": "^10.7.3"
  },
  "peerDependenciesMeta": {
    "mocha": {
      "optional": false
    }
  },
```

And checking the [changes in 9.9.1](https://github.com/eslint/eslint/compare/v9.9.0...v9.9.1), I do see a change to the `.pre-commit-hooks.yaml`, but it only affects the formatting. More relevant is the changes in [package.json](https://github.com/eslint/eslint/compare/v9.9.0...v9.9.1#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519), including:

```diff
-    "@wdio/browser-runner": "^8.40.1",
-    "@wdio/mocha-framework": "^8.40.0",
+    "@wdio/browser-runner": "^9.0.5",
+    "@wdio/mocha-framework": "^9.0.5",
```

The latter brings in Browser-Runner versions with the peer dependency on Mocha 10.x, while the ESLint dependencies under `dev` contain the following which is incompatible:

```
    "mocha": "^8.3.2",
```

## Non-workaround

Given that, I attempted to just pin `@wdio/browser-runner` to `9.0.9`:

```yaml
- repo: https://github.com/eslint/eslint
  rev: v1.11.1
  hooks:
  - id: eslint
    name: Lint and fix JavaScript (ESLint)
    additional_dependencies:
    # Pin to fix dep conflict on install; see eslint/eslint#
    - '@wdio/browser-runner@9.0.9'
```

but unlike all the other times I've done so, it appeared to have no effect on the solve, and 9.11.1's peer dep was still used, and the same error resulted:

```
$ pre-commit run --all eslint
[INFO] Initializing environment for https://github.com/eslint/eslint:@wdio/browser-runner@9.0.9.
[INFO] Installing environment for https://github.com/eslint/eslint.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
An unexpected error has occurred: CalledProcessError: command: ('C:\\Users\\C. A. M. Gerlach\\.cache\\pre-commit\\repotglr1_8v\\node_env-default\\Scripts\\npm.CMD', 'install', '--include=dev', '--include=prod', '--ignore-prepublish', '--no-progress', '--no-save')
return code: 1
stdout: (none)
stderr:
    npm error code ERESOLVE
    npm error ERESOLVE unable to resolve dependency tree
    npm error
    npm error While resolving: eslint@9.11.1
    npm error Found: mocha@8.4.0
    npm error node_modules/mocha
    npm error   dev mocha@"^8.3.2" from the root project
    npm error
    npm error Could not resolve dependency:
    npm error peer mocha@"^10.7.3" from @wdio/browser-runner@9.1.1
    npm error node_modules/@wdio/browser-runner
    npm error   dev @wdio/browser-runner@"^9.0.5" from the root project
    npm error
    npm error Fix the upstream dependency conflict, or retry
    npm error this command with --force or --legacy-peer-deps
    npm error to accept an incorrect (and potentially broken) dependency resolution.

```

## Successful workaround

Therefore, I had to just downgrade `ESLint` to `9.9.0` until a release with the fix is cut:

```yaml
- repo: https://github.com/eslint/eslint
  rev: v9.9.0  # Pin to fix dep conflict on install; see eslint/eslint#
  hooks:
  - id: eslint
```

**Comment by aladdin-add** [comment]
hi, thanks for the report! You've identified two key issues with the dependency management and the release process of the involved packages. Let's break down these issues and discuss them:

### 1. Pre-commit should not install devDependencies

- **Issue**: The `pre-commit` tool is designed to run hooks that are defined in a `.pre-commit-config.yaml` file. When it sets up an environment for a hook, it typically installs the dependencies needed to run the hook. However, by default, it should only install the production dependencies and not the development dependencies (`devDependencies`) as they are not required for the runtime execution of the package.
- **Observation**: If `pre-commit` is installing `wdio` and `mocha` even though they are listed under `devDependencies`, this could be due to the way the `npm install` command is being invoked. By default, `npm install` without any flags will install both production and development dependencies.
- **Solution**:
  - To avoid installing `devDependencies`, you would need to use the `--production` flag with `npm install`.
  - You may want to check if `pre-commit` has an option to exclude `devDependencies` or if you can pass additional arguments to the `npm install` command that `pre-commit` runs.
  - If `pre-commit` does not support this directly, you might have to reach out to the maintainers or look into customizing the setup script to ensure only the necessary dependencies are installed.

### 2. Adding peerDependencies as a breaking change

- **Issue**: According to semantic versioning (semver), a major version bump (e.g., from 8.x.x to 9.x.x) is used when making incompatible API changes, such as adding new `peerDependencies` that could cause conflicts. A minor version (e.g., from 9.0.x to 9.1.x) is intended for backwards-compatible functionality additions, and a patch version (e.g., from 9.1.0 to 9.1.1) is for backwards-compatible bug fixes.
- **Observation**: The addition of a new `peerDependency` in `@wdio/browser-runner` that requires a specific version of `mocha` is indeed a breaking change, as it can prevent the package from being used in environments where the required `mocha` version is not present. This kind of change should have been released in a major version to signal to users that there is a potential for incompatibility.
- **Solution**:
  - consider reporting this to the `@wdio/browser-runner` maintainers, as it's a good practice to follow semver strictly.


On the ESLint side, the version of Mocha has been updated, but the release incorporating this change may take some time: https://github.com/eslint/eslint/issues/18951

**Comment by nzakas** [comment]
Given that we've already updated mocha, is there anything else we need to here @aladdin-add?

**Comment by aladdin-add** [comment]
closing as v9.12.0 has been released.

please feel free to file the issue in wdio; let's see what its maintainers think. /cc @christian-bromann
